dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/PreCheckDataRelationsValidationHook.java

```diff
@@ -146,177 +166,354 @@ public void validateEvent( ValidationErrorReporter reporter, Event event )
+            // since not all AOC, COs exist or the AOC, COs violate programs
+            // default CC rules
+            return;
+        }
+
+        isValid = validateAttributeOptionComboIsInProgramCategoryCombo( reporter, event, program );
+        isValid = validateAttributeCategoryOptionsAreInProgramCategoryCombo( reporter, event ) && isValid;
+        if ( !isValid )
+        {
+            // no need to resolve the AOC id using COs and program CC in case
+            // event.AOC is empty
+            // as we would not find it anyway
+            // no need to cache the AOC id as the payload is invalid
+            return;
+        }
+
+        CategoryOptionCombo aoc = resolveAttributeOptionCombo( reporter, event, program );
+
+        // We should have an AOC by this point. Exit if we do not. The logic of
+        // AOC, COs, CC is complex and there is potential for
+        // missing one of the many cases. Better wrongly invalidate an event
+        // than persisting an invalid event as we
+        // previously did.
+        if ( !validateAttributeOptionComboFound( reporter, event, program, aoc ) )
+            return;
+        if ( !validateAttributeOptionComboMatchesCategoryOptions( reporter, event, program, aoc ) )
+            return;
+
+        // TODO resolving and "caching" the AOC id should move into the preheat
+        // as validations should not access the DB
+        // We "cache" the AOC id for the duration of the import (as the cache is
+        // tied to the context) so the subsequent
+        // EventCategoryOptValidationHook can use the AOC id for validation.
+        // That is necessary if the AOC id is not
+        // provided in the payload and the program cc is non default.
+        reporter.getValidationContext()
+            .cacheEventCategoryOptionCombo( event.getUid(), aoc );
     }
 
-    @Override
-    public void validateRelationship( ValidationErrorReporter reporter, Relationship relationship )
+    private boolean validateAttributeOptionComboExists( ValidationErrorReporter reporter, Event event )
     {
-        validateRelationshipReference( reporter, relationship, relationship.getFrom() );
-        validateRelationshipReference( reporter, relationship, relationship.getTo() );
+        if ( hasNoAttributeOptionComboSet( event ) )
+        {
+            return true;
+        }
+
+        CategoryOptionCombo categoryOptionCombo = reporter.getValidationContext().getBundle().getPreheat()
+            .getCategoryOptionCombo( event.getAttributeOptionCombo() );
+        if ( categoryOptionCombo == null )
+        {
+            reporter.addError( event, E1115, event.getAttributeOptionCombo() );
+            return false;
+        }
+        return true;
     }
 
-    private void validateRelationshipReference( ValidationErrorReporter reporter, Relationship relationship,
-        RelationshipItem item )
+    private boolean hasNoAttributeOptionComboSet( Event event )
```

Comment:
should we rename the method to `hasNoAttributeCategoryOptionSet` instead?

Reply:
`hasNoAttributeOptionComboSet` checks `event.getAttributeOptionCombo()`. Since its not concerned with `AttributeCategoryOption` I would stick with its name.